File: backend/app/models/places.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.database import Base  # correct import for your Base class

class Place(Base):
    __tablename__ = "places"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String, nullable=False)
    description = Column(Text)
    location = Column(String)  # e.g. "Campus Center", "Downtown"
    rating = Column(Float, default=0.0)  # stored average rating
    review_count = Column(Integer, default=0)
    tags = Column(String)  # comma-separated tags like "cafe, study, wifi"
    photo_url = Column(String, nullable=True)
    latitude = Column(Float, nullable=True)
    longitude = Column(Float, nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False, onupdate=lambda: datetime.now(timezone.utc))

    reviews = relationship("PlaceReview", back_populates="place", cascade="all, delete-orphan")


class PlaceReview(Base):
    __tablename__ = "place_reviews"

    id = Column(Integer, primary_key=True, autoincrement=True)
    place_id = Column(Integer, ForeignKey("places.id", ondelete="CASCADE"), nullable=False, index=True)
    reviewer_name = Column(String, nullable=False)
    rating = Column(Float, nullable=False)
    comment = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False)

    place = relationship("Place", back_populates="reviews")

File: frontend/src/services/places.ts
import api from '../lib/api';

export interface Place {
  id: number;
  name: string;
  description?: string | null;
  location?: string | null;
  rating: number;
  review_count: number;
  tags?: string[] | string | null;
  photo_url?: string | null;
  latitude?: number | null;
  longitude?: number | null;
}

export const placeService = {
  list: () => api.get<Place[]>('/places/'),
  top: (limit = 5) => api.get<Place[]>('/places/top', { params: { limit } }),
  nearby: (lat: number, lng: number, radiusKm = 1.5) =>
    api.get<Place[]>('/places/nearby', { params: { lat, lng, radius_km: radiusKm } }),
};
